artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQSession.java

```diff
@@ -302,108 +301,91 @@ public void send(final ProducerInfo producerInfo,
+                  if (result == RoutingStatus.NO_BINDINGS && dest.isQueue()) {
+                     throw new InvalidDestinationException("Cannot publish to a non-existent Destination: " + dest);
+                  }
+               } catch (Exception e) {
+                  if (anyException[0] == null) {
+                     anyException[0] = e;
+                  }
+               }
+               if (count.decrementAndGet() == 0) {
+                  if (anyException[0] != null) {
+                     ActiveMQServerLogger.LOGGER.warn(anyException[0].getMessage(), anyException[0]);
+                     connection.sendException(anyException[0]);
+                  } else {
+                     if (sendProducerAck) {
+                        try {
+                           ProducerAck ack = new ProducerAck(producerInfo.getProducerId(), messageSend.getSize());
+                           connection.dispatchAsync(ack);
+                        } catch (Exception e) {
+                           ActiveMQServerLogger.LOGGER.warn(e.getMessage(), e);
+                           connection.sendException(e);
+                        }
+                     } else {
+                        try {
+                           //maybe use this: connection.getContext().setDontSendReponse(false);
+                           Response response = new Response();
+                           response.setCorrelationId(messageSend.getCommandId());
+                           connection.dispatchAsync(response);
+                        } catch (Exception e) {
+                           ActiveMQServerLogger.LOGGER.warn(e.getMessage(), e);
+                           connection.sendException(e);
+                        }
+                     }
+                  }
+               }
+            })) {
+               this.connection.getContext().setDontSendReponse(false);
+               throw new ResourceAllocationException("Queue is full " + address);
+            }
          } else {
-            coreMsg.putByteProperty(org.apache.activemq.artemis.api.core.Message.HDR_ROUTING_TYPE, RoutingType.MULTICAST.getType());
-         }
-         RoutingStatus result = getCoreSession().send(coreMsg, false, actualDestinations[i].isTemporary());
+            //non-persistent messages goes here, by default we block on the calling thread
+            //which means the whole connection is blocked when address is full.
+            final CountDownLatch latch = new CountDownLatch(1);
+            if (!store.checkMemory(() -> {
+               latch.countDown();
+            })) {
+               throw new ResourceAllocationException("Queue is full " + address);
+            }
 
-         if (result == RoutingStatus.NO_BINDINGS && actualDestinations[i].isQueue()) {
-            throw new InvalidDestinationException("Cannot publish to a non-existent Destination: " + actualDestinations[i]);
-         }
+            latch.await();
```

Comment:
The session was playing with setAutoRead from Netty, disabling reads.. instead of blocking a thread. That's how this is supposed to work.


If there's anything wrong with setAutoRead.. fix it.. don't remove it.